src/Lib.P2P.Tests/PubSub/NotificationServiceTest .cs

```diff
@@ -0,0 +1,240 @@
+                Assert.AreEqual(3ul, ns.MesssagesPublished);
+                Assert.AreEqual(3ul, ns.MesssagesReceived);
+            }
+            finally
+            {
+                await ns.StopAsync();
+            }
+        }
+
+        [TestMethod]
+        public async Task Topics()
+        {
+            var ns = new PubSubService {LocalPeer = self};
+            await ns.StartAsync();
+            try
+            {
+                var topicA = Guid.NewGuid().ToString();
+                var topicB = Guid.NewGuid().ToString();
+                var csA = new CancellationTokenSource();
+                var csB = new CancellationTokenSource();
+
+                await ns.SubscribeAsync(topicA, msg => { }, csA.Token);
+                await ns.SubscribeAsync(topicA, msg => { }, csA.Token);
+                await ns.SubscribeAsync(topicB, msg => { }, csB.Token);
+
+                var topics = (await ns.SubscribedTopicsAsync(csA.Token)).ToArray();
+                Assert.AreEqual(2, topics.Length);
+                CollectionAssert.Contains(topics, topicA);
+                CollectionAssert.Contains(topics, topicB);
+
+                csA.Cancel();
+                topics = (await ns.SubscribedTopicsAsync(csA.Token)).ToArray();
+                Assert.AreEqual(1, topics.Length);
+                CollectionAssert.Contains(topics, topicB);
+
+                csB.Cancel();
+                topics = (await ns.SubscribedTopicsAsync(csA.Token)).ToArray();
+                Assert.AreEqual(0, topics.Length);
+            }
+            finally
+            {
+                await ns.StopAsync();
+            }
+        }
+
+        [TestMethod]
+        public async Task Subscribe()
+        {
+            var ns = new PubSubService {LocalPeer = self};
+            await ns.StartAsync();
+            try
+            {
+                var topic = Guid.NewGuid().ToString();
+                var cs = new CancellationTokenSource();
+                var messageCount = 0;
+                await ns.SubscribeAsync(topic, msg => { ++messageCount; }, cs.Token);
+                await ns.SubscribeAsync(topic, msg => { ++messageCount; }, cs.Token);
+
+                await ns.PublishAsync(topic, "", cs.Token);
+                Assert.AreEqual(2, messageCount);
+            }
+            finally
+            {
+                await ns.StopAsync();
+            }
+        }
+
+        [TestMethod]
+        public async Task Subscribe_HandlerExceptionIsIgnored()
+        {
+            var ns = new PubSubService {LocalPeer = self};
+            await ns.StartAsync();
+            try
+            {
+                var topic = Guid.NewGuid().ToString();
+                var cs = new CancellationTokenSource();
+                var messageCount = 0;
+                await ns.SubscribeAsync(topic, msg =>
+                {
+                    ++messageCount;
+                    throw new Exception();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)